platforms/core-configuration/configuration-cache/src/integTest/groovy/org/gradle/configurationcache/isolated/IsolatedProjectsToolingApiParallelModelQueryIntegrationTest.groovy

```diff
@@ -0,0 +1,84 @@
+/*
+ * Copyright 2024 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.configurationcache.isolated
+
+import org.gradle.configurationcache.fixtures.SomeToolingModel
+import org.gradle.test.fixtures.file.TestFile
+import org.gradle.test.fixtures.server.http.BlockingHttpServer
+import org.junit.Rule
+
+class IsolatedProjectsToolingApiParallelModelQueryIntegrationTest extends AbstractIsolatedProjectsToolingApiIntegrationTest {
+    @Rule
+    BlockingHttpServer server = new BlockingHttpServer()
+
+    def setup() {
+        server.start()
+    }
+
+    def "intermediate model is cached and reused for nested concurrent requests"() {
+        withSomeToolingModelBuilderPluginInBuildSrc("""
+            Thread.sleep(3000) // Simulate long-running builder to ensure overlap
```

Comment:
Do we need this call to `sleep()`? It's usually pretty unreliable. The call to the blocking URL in the build action can move into the model builder instead to ensure the models are being built in parallel.

Reply:
The individual models are not built in parallel due to a project lock. That's why I am catching it with the `sleep`.
See the other comment for details.